tests/usm/usm_api.h

```diff
@@ -0,0 +1,688 @@
+ *          check the method results (target pointer for the copy operations for
+ *          example)
+ */
+template <typename T, size_t count, template <typename, size_t> class checkT,
+          typename caller, allocation alloc, size_t numEvents>
+struct test {
+  static void run(sycl_cts::util::logger &log, const std::string &typeName) {
+    using verifier_t = checkT<T, count>;
+    using storage_t = storage<T, count, alloc>;
+    using events_t = std::vector<sycl::event>;
+    // Elements to process in event_generator. Large number is selected to make
+    // kernel execution longer if numEvents > 0
+    constexpr size_t gen_buf_size = gen_buf_size_selector<numEvents>::value;
+    using gen_arr_t = std::array<event_generator<T, gen_buf_size>, numEvents>;
+
+    log.debug([&] {
+      std::string message("Running test for ");
+      message += verifier_t::template description<alloc>() + " for ";
+      message += type_name_string<T>::get(typeName) + " ...";
+      return message;
+    });
+
+    try {
+      auto queue = sycl_cts::util::get_cts_object::queue();
+
+      if (!check_device_support<alloc>(log, queue)) return;
+
+      log.debug("... allocate USM memory storage");
+
+      auto scopedStorage = storage_t::get(queue, T{0});
+      const auto ptr = scopedStorage.get();
+
+      log.debug("... prepare specific verifier instance");
+      verifier_t verifier(queue, 1);
+
+      // Event generators array
+      gen_arr_t gens{};
+      // Non-zero (non default) value to initialize event generators source
+      // buffers
+      T init_value{1};
+
+      auto action = [&](typename caller::type &parent) {
+        if constexpr (numEvents == 0) {
+          log.debug("... submit USM operation with 0 events");
+          verifier.call(parent, ptr);
+        } else {
+          static_assert(std::is_same_v<typename caller::type, sycl::queue>,
+                        "Tests with events are not acceptible for handler");
+          // List of events that tested usm operation should wait
+          events_t depEvents(numEvents);
+          // Run time-consuming tasks to generate events
+          for (size_t i = 0; i < depEvents.size(); ++i) {
+            depEvents[i] = gens[i].init(parent, init_value);
+          }
+
+          if constexpr (numEvents == 1) {
+            auto event = verifier.call(parent, ptr, depEvents[0]);
+            event.wait();
+          } else {
+            auto event = verifier.call(parent, ptr, depEvents);
+            event.wait();
+          }
+
+          // Fast copy arrays for future check. If tested function does not
+          // wait for the passed events to complete, then an incompletely
+          // initialized 'src_arr' will be copied to 'dst_arr' and the test
+          // will fail during validation. Reverse order is used to increase
+          // probability of data race.
+          for (size_t i = numEvents; i > 0; --i) {
+            const size_t idx = i - 1;
```

Comment:
```suggestion
          for (size_t i = numEvents - 1; i + 1 > 0; --i) {
```
Using signed numbers would be easier

Reply:
I've applied this comment in this commit: 8e176b5cfac40645d010a79e266ad37397f4c762